### ISSUE TITLE
Truncate the length of variable names

### ISSUE THREAD

**Issue by statzhero** [sfirke/janitor#332]
## Feature requests

Just to confirm: there's no option yet to truncate variable names? 

E.g. `clean_names(x, max = 24)`

**Comment by sfirke** [comment]
There is not.  I am open to adding this, either as a separate function or as part of `clean_names`.  

Would you shorten it by truncating from the end, then forcing uniqueness?  I sometimes have very long survey questions with identical beginnings and in those cases I take maybe the first and last 10 characters, separated by a `_`.

**Comment by statzhero** [comment]
> shorten it by truncating from the end, then forcing uniqueness
Yes, that sounds most appealing.  `base::abbreviate` has also some interesting features.

**Comment by jzadra** [comment]
I second `abbreviate()`. I use it for exactly this purpose and it has highly configurable arguments.

**Comment by sfirke** [comment]
Is this the same as #201 ?  That's a couple of votes then.  I wasn't aware of `base::abbreviate` but it looks like it's been missing from my life.

I like the idea but this would be adding a bunch of new arguments to `clean_names` to control the truncation in `abbreviate`.  `abbreviate` seems pretty good already, if I would just build on that then maybe an immediate solution is to add a link to `abbreviate` to the clean_names documentation?  I think most people can write the extra line to follow `clean_names` with `abbreviate`, the thing is knowing it exists.

**Comment by billdenney** [comment]
I'll try to pull something like this into the #340 rewrite.

**Comment by sfirke** [comment]
If you think it doesn't add too much for something that already exists in a
modular form in abbreviate.

Maybe just a max length argument and in the docs note that finer control of
truncation can be achieved through abbreviate?

**Comment by billdenney** [comment]
I realized that I missed this feature in the `clean_names()` rewrite.  What is the advantage of `abbreviate()` over simply using `substr()`?  Also, `abbreviate()` appears not to always actually shorten the text.  So, if you were looking for something with <=2 characters, you may not get that.

**Comment by sfirke** [comment]
I'm not finding abbreviate useful for my cases where the variable is a long text string from a survey.

`x <- "Please indicate your level of agreement with the following statement: my manager takes actions to make me feel valued."`

This is not useful:
```
abbreviate(x, minlength = 30, method = "both.sides")
> "Plsiyloawtfsttmnt:mmtatmmfvld."
```

I much prefer:
```
> paste(stringr::str_sub(x, 1, 15), stringr::str_sub(x, start = -15), sep = "_")
[1] "Please indicate_me feel valued."
```
As a better use of 30 characters.  I may be biased by this being my main use case for this problem.

**Comment by jzadra** [comment]
I agree that a truncated version is more useful than a long abbreviation.  I played around with `abbreviate()` a little and couldn't get anything much better. 

One thing that occurred to me: are there special things to watch for that are more critical pieces of info in a long colname that we want to try to keep?  For instance numerals - "Please indicate your level in 2019", "Please indicate your level in 2020", or capitals, etc?

If so, we could check for these markers and make sure they are always kept.

Another thing would be to combine truncation with abbreviation for a string above some number of chars - truncate the first 15 or 20 keeping the special markers, then abbreviate the rest?

**Comment by billdenney** [comment]
@jzadra, I was thinking that if we went down a path with a lot of controls for how abbreviation happens, then we may want another pair of functions (e.g. `make_abbrev_names()` and `abbrev_names()`, "abbrev" is specifically abbreviated because that makes me smile).  Then, you could run:

```r
data %>%
  abbrev_names([all the controls for abbreviation]) %>%
  clean_names([all the controls for cleaning])
```

In my mind, the part that fits in `make_clean_names()` (and therefore `clean_names()`) is the part that @sfirke suggested:  Keep the beginning and end of the name with an underscore in the middle.  In general, I think that the beginning and end of the column name contain most of the more critical information.  (In your example, "2019" and "2020" are both at the end, and the "Please indicate" would give clarity to a column class.)

One other item that I'd suggest is that abbreviated names would not be *guaranteed* to have the number of characters due to duplicated column names having `_1` and counting up from there appended.  The guarantee would be that before de-duplication, the number of characters would be the abbreviated number.  Does that seem reasonable?

For my use cases, I find myself often doing the following:

```r
data %>%
  clean_names() %>%
  rename(
    # make the names what I actually want them to be, but
    # start from something known to be ok and unique
  )
```

**Comment by jzadra** [comment]
That makes sense to me.  And I often do something similar to your use case - let clean_names() get them most of the way there and unique, and then modify what needs to be changed manually.
